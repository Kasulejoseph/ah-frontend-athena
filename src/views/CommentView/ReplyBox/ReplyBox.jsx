import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { CommentAction } from '../../../actions/CommentActions/CommentAction';
import {
  CommentEditAction,
  CommentDeleteAction
} from '../../../actions/CommentActions/CommentEditAction';
import {
  ReplyPostAction,
  CommentGetAction
} from '../../../actions/CommentActions/CommentGetAction';
import CommentCard from '../../../components/Comments/CommentCard/CommentCard';
import ModalPage from '../../../components/Likes/LoginModal';

export class ReplyBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = { modal: false };
  }

  onSubmitReply = id => {
    const commentobj = this.state;

    const {
      ReplyPostAction,
      CommentEditAction,
      CommentAction,
      art_slug,
      isEdit
    } = this.props;
    const comment = {
      comment_body: commentobj.comment_body
    };

    if (isEdit) {
      if (id === null) {
        CommentAction(comment, art_slug);
      } else {
        CommentEditAction(id, art_slug, comment);
      }
    } else {
      const { parentId, art_slug } = this.props;
      if (parentId && art_slug) {
        ReplyPostAction(id, comment, art_slug);
      } else {
        CommentAction(comment, art_slug);
      }
    }
  };

  checkToken = () => {
    const token = window.localStorage.getItem('token');
    if (token) {
      this.setState({
        modal: false
      });
    } else {
      this.setState({
        modal: true
      });
      return;
    }
  };

  toggle = () => {
    const { modal } = this.state;
    this.setState({
      modal: !modal
    });
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };
  onSubmit = e => {
    e.preventDefault();
    this.checkToken();
    const { childId } = this.props;
    this.onSubmitReply(childId);
  };
  render() {
    const { childId, history } = this.props;
    const { modal } = this.state;
    const url = window.location.pathname;
    return (
      <div>
        <CommentCard
          childId={childId}
          onChange={this.onChange}
          onSubmit={this.onSubmit}
          onSubmitReply={this.onSubmit}
        />
        <ModalPage
          title="Please login before you can rate the article"
          modal={modal}
          toggle={this.toggle}
          fallback={url}
          history={history}
          md="12"
        />
      </div>
    );
  }
}

ReplyBox.propTypes = {
  CommentAction: PropTypes.func.isRequired,
  history: PropTypes.func.isRequired,
  ReplyPostAction: PropTypes.func.isRequired,
  CommentEditAction: PropTypes.func.isRequired,
  art_slug: PropTypes.string,
  isEdit: PropTypes.bool,
  parentId: PropTypes.number,
  childId: PropTypes.number
};
ReplyBox.defaultProps = {
  art_slug: '',
  isEdit: false,
  parentId: null,
  childId: null
};
export const mapStateToProps = state => {
  return {
    currentUser: state.login.login,
    art_slug: state.articles.view_article.slug
  };
};

export default connect(
  mapStateToProps,
  {
    CommentAction,
    ReplyPostAction,
    CommentGetAction,
    CommentEditAction,
    CommentDeleteAction
  }
)(ReplyBox);
