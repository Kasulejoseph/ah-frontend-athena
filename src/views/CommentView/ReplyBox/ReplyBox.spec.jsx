import React from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { ReplyBox, mapStateToProps } from './ReplyBox';

const initialState = {
  comment: {
    commentList: {},
    comment: {},
    replyList: [],
    refresh: false
  },
  login: {
    login: {
      username: 'kasule',
      email: 'okello@andela.com',
      password: '',
      errors: ''
    }
  },
  isEdit: false,
  articles: {
    view_article: {
      slug: 'someslug'
    }
  }
};
let store;
const mockStore = configureStore();
store = mockStore(initialState);
let wrapper;
const defaultProps = {
  isEdit: false,
  CommentAction: jest.fn(),
  CommentEditAction: jest.fn(),
  ReplyPostAction: jest.fn(),
  onSubmitReply: jest.fn(),
  onSubmit: jest.fn(),
  onChange: jest.fn(),
  art_slug: 'mocked-slug',
  parentId: '1'
};

const updateProps = props => Object.assign(defaultProps, props);
describe('test ReplyBox view', () => {
  beforeEach(() => {
    wrapper = shallow(<ReplyBox {...defaultProps} />);
  });
  it('should  ReplyBox match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('test handle onSubmitReply event', () => {
    defaultProps.isEdit = true;
    wrapper = shallow(<ReplyBox {...defaultProps} />);
    jest.setTimeout(10000);
    wrapper.instance().onSubmitReply(null);
    expect(defaultProps.CommentAction).toBeCalled();
  });

  it('should handle onchange function', () => {
    wrapper
      .instance()
      .onChange({ target: { name: 'comment', value: 'my comment' } });
    expect(wrapper.state('comment')).toEqual('my comment');
  });

  it('should handle onsubmit function', () => {
    wrapper.instance().onSubmit({ preventDefault() {} });
    expect(wrapper.state()).toEqual({ modal: true });
  });

  it('should handle onSubmitReply function', (id = jest.fn()) => {
    wrapper.instance().onSubmitReply(id);
    wrapper.setState({
      isEdit: true
    });
    expect(wrapper.state('isEdit')).toEqual(true);
  });

  it('should recieve props', () => {
    const wrapper2 = mount(
      <Provider store={store}>
        <ReplyBox {...defaultProps} />
      </Provider>
    );
    wrapper2.setState({
      isEdit: true
    });
    expect(defaultProps.CommentAction).toBeCalled();
    expect(wrapper2.state('isEdit')).toEqual(true);
  });

  it('should map state to props', () => {
    expect(mapStateToProps(initialState).currentUser['username']).toEqual(
      'kasule'
    );
  });

  it('should test prototype of data', () => {
    const wrapper2 = mount(
      <Provider store={store}>
        <ReplyBox {...updateProps} />
      </Provider>
    );
    expect(wrapper2.instance().props.children).toEqual(<ReplyBox />);
  });

  it('test parentID and slug in props', () => {
    defaultProps.isEdit = false;
    wrapper = shallow(<ReplyBox {...defaultProps} />);
    wrapper.instance().onSubmitReply(null);
    expect(defaultProps.ReplyPostAction).toBeCalled();
  });

  it('test parentID and slug in props', () => {
    defaultProps.isEdit = false;
    defaultProps.art_slug = undefined;
    wrapper = shallow(<ReplyBox {...defaultProps} />);
    wrapper.instance().onSubmitReply(null);
    expect(defaultProps.CommentAction).toBeCalled();
  });
});
