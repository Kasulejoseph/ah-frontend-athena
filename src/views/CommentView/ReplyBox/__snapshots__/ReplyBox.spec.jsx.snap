// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ReplyBox view should  ReplyBox match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReplyBox
    CommentAction={[MockFunction]}
    CommentEditAction={[MockFunction]}
    ReplyPostAction={[MockFunction]}
    art_slug="mocked-slug"
    childId={null}
    isEdit={false}
    onChange={[MockFunction]}
    onSubmit={[MockFunction]}
    onSubmitReply={[MockFunction]}
    parentId="1"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <CommentCard
          childId={null}
          onChange={[Function]}
          onSubmit={[Function]}
          onSubmitReply={[Function]}
        />,
        <ModalPage
          fallback="/"
          md="12"
          modal={false}
          title="Please login before you can rate the article"
          toggle={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "childId": null,
          "onChange": [Function],
          "onSubmit": [Function],
          "onSubmitReply": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fallback": "/",
          "history": undefined,
          "md": "12",
          "modal": false,
          "title": "Please login before you can rate the article",
          "toggle": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <CommentCard
            childId={null}
            onChange={[Function]}
            onSubmit={[Function]}
            onSubmitReply={[Function]}
          />,
          <ModalPage
            fallback="/"
            md="12"
            modal={false}
            title="Please login before you can rate the article"
            toggle={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "childId": null,
            "onChange": [Function],
            "onSubmit": [Function],
            "onSubmitReply": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fallback": "/",
            "history": undefined,
            "md": "12",
            "modal": false,
            "title": "Please login before you can rate the article",
            "toggle": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
